File: python/pv_single_diode_model.py
"""
pv_single_diode_model.py — 光伏电池单二极管模型，绘制 I-V 与 P-V 曲线
对应 pv_single_diode_model.m：计算并绘制 I-V/P-V 特性，输出最大功率点。
运行：uv run python pv_single_diode_model.py
"""
import numpy as np
import plot

from pv_model import pv_params, pv_current

pv = pv_params()
V_oc = pv["Voc_stc"]

V_PV = np.linspace(0, V_oc, 1000)
I_PV = pv_current(V_PV, pv)
P = V_PV * I_PV

i_max = int(np.argmax(P))
V_mppt, P_mppt = V_PV[i_max], P[i_max]

print(f"开路电压: {V_oc:.2f} V, 短路电流: {pv['Isc_stc']:.2f} A")
print(f"最大功率点: V={V_mppt:.2f} V, P={P_mppt:.2f} W")

# I-V 特性曲线
fig, ax = plot.new_figure("pv_single_diode_iv")
ax.plot(V_PV, I_PV, "-b", label="I-V")
ax.set_xlabel("Voltage (V)")
ax.set_ylabel("Current (A)")
ax.set_title("I-V 特性曲线")
ax.legend()
plot.save(fig)

# P-V 特性曲线
fig, ax = plot.new_figure("pv_single_diode_pv")
ax.plot(V_PV, P, "-r", label="P-V")
ax.plot(V_mppt, P_mppt, "ko", markersize=8, markerfacecolor="k", label="MPP")
ax.set_xlabel("Voltage (V)")
ax.set_ylabel("Power (W)")
ax.set_title("P-V 特性曲线")
ax.legend()
plot.save(fig)

File: 代码/pv_single_diode_model.m
function [V_PV, I_PV] = pv_single_diode_model()
% 光伏电池单二极管模型函数
% 功能：计算并绘制光伏电池的电流-电压(I-V)和功率-电压(P-V)特性曲线
% 输出：V_PV - 电压向量，I_PV - 电流向量
% 运行环境：MATLAB 或 Octave
% 用法：直接调用 pv_single_diode_model 即可看到曲线和最大功率结果。
%       光伏模型参数统一来自 pv_params() + pv_current()。

% ==== 光伏组件参数（统一参数源） ====
pv = pv_params();
V_oc = pv.Voc_stc;

% ==== 生成电压采样点并计算 I-V/P-V 特性 ====
V_PV = linspace(0, V_oc, 1000);
I_PV = pv_current(V_PV, pv);   % 统一单二极管模型
P = V_PV .* I_PV;               % 输出功率

% 最大功率点
[P_mppt, imax] = max(P);
V_mppt = V_PV(imax);

fprintf('开路电压: %.2f V, 短路电流: %.2f A\n', V_oc, pv.Isc_stc);
fprintf('最大功率点: V=%.2f V, P=%.2f W\n', V_mppt, P_mppt);

% ==== 结果可视化 ====
% I-V 特性曲线
figure;
plot(V_PV, I_PV, '-b');
xlabel('Voltage (V)');
ylabel('Current (A)');
title('I-V 特性曲线');
grid on;

% P-V 特性曲线
figure;
plot(V_PV, P, '-r');
xlabel('Voltage (V)');
ylabel('Power (W)');
title('P-V 特性曲线');
grid on;
hold on;
plot(V_mppt, P_mppt, 'ko', 'MarkerSize', 8, 'MarkerFaceColor', 'k');  % 标记MPP

end